sdk/helper/storagepacker/storagepacker_v2.go

```diff
@@ -0,0 +1,799 @@
+	s.bucketsCacheLock.RUnlock()
+
+	if found {
+		bucket = bucketRaw.(*LockedBucket)
+	} else {
+		// Read from underlying view
+		storageEntry, err := s.BucketStorageView.Get(ctx, bucketKey)
+		if err != nil {
+			return errwrap.Wrapf("failed to read packed storage value: {{err}}", err)
+		}
+		if storageEntry == nil {
+			return nil
+		}
+
+		bucket, err = s.DecodeBucket(storageEntry)
+		if err != nil {
+			return errwrap.Wrapf("error decoding existing storage entry for deletion: {{err}}", err)
+		}
+
+		bucket.LockEntry = lock
+
+		s.bucketsCacheLock.Lock()
+		s.bucketsCache.Insert(cacheKey, bucket)
+		s.bucketsCacheLock.Unlock()
+	}
+
+	if len(bucket.ItemMap) == 0 {
+		return nil
+	}
+
+	itemHash := GetItemIDHash(itemID)
+
+	_, ok := bucket.ItemMap[itemHash]
+	if !ok {
+		return nil
+	}
+
+	delete(bucket.ItemMap, itemHash)
+	return s.storeBucket(ctx, bucket, true)
+}
+
+// GetItem fetches the storage entry for a given key from its corresponding
+// bucket.
+func (s *StoragePackerV2) GetItem(ctx context.Context, itemID string) (*Item, error) {
+	if itemID == "" {
+		return nil, fmt.Errorf("empty item ID")
+	}
+
+	bucketKey := s.BucketStorageKeyForItemID(itemID)
+	cacheKey := s.GetCacheKey(bucketKey)
+
+	lock := locksutil.LockForKey(s.storageLocks, cacheKey)
+	lock.RLock()
+	defer lock.RUnlock()
+
+	var bucket *LockedBucket
+
+	s.bucketsCacheLock.RLock()
+	_, bucketRaw, found := s.bucketsCache.LongestPrefix(cacheKey)
+	s.bucketsCacheLock.RUnlock()
+
+	if found {
+		bucket = bucketRaw.(*LockedBucket)
+	} else {
+		// Read from underlying view
+		storageEntry, err := s.BucketStorageView.Get(ctx, bucketKey)
+		if err != nil {
+			return nil, errwrap.Wrapf("failed to read packed storage value: {{err}}", err)
+		}
+		if storageEntry == nil {
+			return nil, nil
+		}
+
+		bucket, err = s.DecodeBucket(storageEntry)
+		if err != nil {
+			return nil, errwrap.Wrapf("error decoding existing storage entry: {{err}}", err)
+		}
+
+		bucket.LockEntry = lock
+
+		s.bucketsCacheLock.Lock()
+		s.bucketsCache.Insert(cacheKey, bucket)
+		s.bucketsCacheLock.Unlock()
+	}
+
+	if len(bucket.ItemMap) == 0 {
+		return nil, nil
+	}
+
+	itemHash := GetItemIDHash(itemID)
+
+	data, ok := bucket.ItemMap[itemHash]
+	if !ok {
+		return nil, nil
+	}
+
+	return &Item{
+		ID:   itemID,
+		Data: data,
+	}, nil
+}
+
+// PutItem stores a storage entry in its corresponding bucket
+func (s *StoragePackerV2) PutItem(ctx context.Context, item *Item) error {
+	if item == nil {
+		return fmt.Errorf("nil item")
+	}
+
+	if item.ID == "" {
+		return fmt.Errorf("missing ID in item")
+	}
+
+	if item.Data == nil {
+		return fmt.Errorf("missing data in item")
+	}
+
+	if item.Message != nil {
+		return fmt.Errorf("'Message' is deprecated; use 'Data' instead")
+	}
+
+	// Get the bucket key
+	bucketKey := s.BucketStorageKeyForItemID(item.ID)
+	cacheKey := s.GetCacheKey(bucketKey)
+
+	lock := locksutil.LockForKey(s.storageLocks, cacheKey)
+	lock.Lock()
```

Comment:
I think there's a second concurrency problem here (and probably in the Get and Delete case too?)

The lock that we get is based on BucketStorageKeyForItemID, but another operation might have re-sharded between the time we consulted in the bucket cache for the longest prefix and the time we acquire the lock.  So, we may have gotten the lock for 00 but the call to LongestPrefix below (our second on the same item, essentially) will get the bucket 00/0.

Because the lock is for the wrong bucket, we might end up putting this item into the wrong bucket as well, if we get particularly unlucky.  If somebody re-shards again after calling LongestPrefix, we might have a reference to bucket 00/0 but we should have put the key into bucket 00/0/0.